core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -71,6 +78,57 @@ public static CloseableIterable<FileScanTask> splitFiles(
     return CloseableIterable.combine(splitTasks, tasks);
   }
 
+  public static <T extends FileScanTask> Iterable<CombinedScanTask> planTasks(
+      List<T> splitFiles,
+      long splitSize,
+      int lookback,
+      long openFileCost,
+      StructProjection projection) {
+    Preconditions.checkArgument(splitSize > 0, "Invalid split size (negative or 0): %s", splitSize);
+    Preconditions.checkArgument(
+        lookback > 0, "Invalid split planning lookback (negative or 0): %s", lookback);
+    Preconditions.checkArgument(
+        openFileCost >= 0, "Invalid file open cost (negative): %s", openFileCost);
+
+    // Check the size of delete file as well to avoid unbalanced bin-packing
+    Function<FileScanTask, Long> weightFunc =
+        file ->
+            Math.max(
+                file.length()
+                    + file.deletes().stream().mapToLong(ContentFile::fileSizeInBytes).sum(),
+                (1 + file.deletes().size()) * openFileCost);
+
+    ListMultimap<StructLikeWrapper, FileScanTask> groupedFiles =
+        Multimaps.newListMultimap(Maps.newHashMap(), Lists::newArrayList);
+
+    for (T task : splitFiles) {
+      PartitionSpec spec = task.spec();
+      StructProjection projectedStruct =
+          StructProjection.create(spec.partitionType(), projection.type());
+      Types.StructType projectedPartitionType = projectedStruct.type();
+      StructLikeWrapper wrapper =
+          StructLikeWrapper.forType(projectedPartitionType)
```

Comment:
This invocation is very expensive as `forType` would call this constructor:

```
private StructLikeWrapper(Types.StructType type) {
  this(Comparators.forType(type), JavaHash.forType(type));
}
```

Instead, we should try to call `forType` only once and then use `copyFor`.

```
StructLikeWrapper wrapper = StructLikeWrapper.forType(...);

for (T task : splitFiles) {
  StructProjection projection = ...
  projection.wrap(task.file().partition());
  groupedFiles.put(wrapper.copyFor(projection.copy()), task);
}
```

Reply:
Done